src/basic/fd-util.c

```diff
@@ -822,6 +822,32 @@ int fd_reopen(int fd, int flags) {
         return new_fd;
 }
 
+int fd_reopen_propagate_flags(int fd, int flags, int propagate_flags) {
+        /* Invokes fd_reopen(fd, flags),
+         * but propagates selected flags if set on the original fd.
+         *
+         * You should use this for example if the original fd potentially is O_APPEND,
+         * otherwise we get rather "unexpected" behavior.
+         * Unless you intentionally want to overwrite pre-existing data,
+         * and have your output overwritten by the next guy.
+         *
+         * Used for "systemd-run --pty >> some-log", but may be used
+         * to propagate other flags as well, if there is a use case. */
+
+        /* Whitelist supported flags here.
+         * For now, the only supported flag is O_APPEND. */
+        const int supported_flags_mask = O_APPEND;
+        assert((propagate_flags & ~supported_flags_mask) == 0);
+
+        int orig_status_flags = fcntl(fd, F_GETFL);
+        if (orig_status_flags == -1)
+                return negative_errno();
```

Comment:
we generally check for errors of syscalls with < 0, please do the same here. And negative_errno() has a special use, which doesn't apply here. Please just do "return -errno".